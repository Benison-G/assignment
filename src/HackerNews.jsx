import React, { useEffect, useState } from 'react';
import Axios from 'axios';
import './TableRows.scss';
import Graph from './Graph';

/**
 * Component to render the Front Page of the application
 */
function HackerNews() {
    const [items, setItems] = useState([]);

    /**
     * Makes api call to retrieve the items to display on front page.
     * Will be executed only once on initial render(once the component mounts)
     */
    useEffect(() => {
        // Get the items from the localStorage if the component is refreshed otherwise call API
        const storedData = JSON.parse(localStorage.getItem("items"));
        if(!storedData) {
            Axios.get('https://hn.algolia.com/api/v1/search?tags=front_page').then(response => {
                const serviceResponse = response.data.hits.map(obj=> ({ ...obj, voteCount: 0 }));
                localStorage.setItem("items", JSON.stringify(serviceResponse));
                setItems(serviceResponse);
            });
        } else {
            setItems(storedData);
        }
    }, []);

    // Updates the localStorage on every state change.
    useEffect(() => {
        localStorage.setItem("items", JSON.stringify(items));
    })

    /**
     * Increments the vote count
     * @param {string} id The unique identifier of the list item(object);
     */
    function upVote(id) {
        const elementsIndex = items.findIndex(element => element.objectID === id )
        items[elementsIndex].voteCount++;
        // one of its values but it's still the same array, so React will not re-render as the state hasn't changed. 
        // The new array is still the old array. So using spread operator to avoid this behaviour.
        setItems([...items]);
    }

    return (
        <table cellPadding="0" cellSpacing="0" width="100%" background-color="#f6f6ef">
         <tr>
            <td>
               <table border="0" cellPadding="0" cellSpacing="0" width="100%" style={{ padding:"2px" }}>
                    <tr bgcolor="#ff6600" style={{ color:"white", fontWeight: "bold" }}>
                        <td> Comments </td>
                        <td> Vote Count</td>
                        <td> Upvote</td>
                        <td> News Details</td>
                    </tr>
                    {
                        items.map((item) => {
                            return (
                            <tr className="table-rows" key={item.objectID}>
                                <td>{ item.num_comments }</td>
                                <td>{ item.voteCount }</td>
                                <td>
                                    <div 
                                        style={{ 
                                            width: 0, 
                                            height: 0, 
                                            borderTop: 0, 
                                            borderLeft: "5px solid transparent", 
                                            borderRight: "5px solid transparent", 
                                            borderBottom: "10px solid grey" 
                                        }}
                                        onClick={() => upVote(item.objectID)} 
                                    />
                                </td>
                                <td>{ item.title }</td>
                            </tr>)
                        })
                    }
               </table>
            </td>
         </tr>
         <tr>
            <td style={{ color: "#ff6600", fontWeight: "bold", padding: "5px" }}>
                 <div style={{ float: "right" }}><span>Previous&nbsp;</span>|<span>&nbsp;Next</span></div>
            </td>
         </tr>
         <tr>
            <td style={{ borderTop: "2px solid #ff6600", borderBottom: "4px solid #ff6600" }}>
                {/* <Graph data={getGraphPoints(items)}/> */}
            </td>
         </tr>
      </table>
    );
}

function getGraphPoints(items) {
    const points = [];
    items.map(item => {
        const obj = {}
        obj.x = item.objectID;
        obj.y = item.voteCount;
        points.push(obj);
    })
    return points;
}

export default HackerNews;