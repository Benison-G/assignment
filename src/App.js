import React from "react";
import HackerNews from "./HackerNews";


function App() {
  return (
   <HackerNews />
  );
}

export default App;
