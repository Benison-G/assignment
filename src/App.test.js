import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';
import HackerNews from './HackerNews';

test('app component render', () => {
  const tree = renderer
    .create(<App />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('HackerNews component render', () => {
  const items = [
    {
      title: 'my post',
      objectID: '123',
      num_comments: 20,
      voteCount: 20
    }
  ]
  const tree = renderer
    .create(<HackerNews items={items} upVote={jest.fn()}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
