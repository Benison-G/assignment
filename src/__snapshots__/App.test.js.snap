// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HackerNews component render 1`] = `
<table>
  <tr>
    <td>
      <table>
        <tr
          bgcolor="#ff6600"
          style={
            Object {
              "color": "white",
              "fontWeight": "bold",
            }
          }
        >
          <td>
             Comments 
          </td>
          <td>
             Vote Count
          </td>
          <td>
             Upvote
          </td>
          <td>
             News Details
          </td>
        </tr>
        <tr
          className="table-rows"
        >
          <td>
            20
          </td>
          <td>
            20
          </td>
          <td>
            <div
              onClick={[Function]}
              style={
                Object {
                  "borderBottom": "10px solid grey",
                  "borderLeft": "5px solid transparent",
                  "borderRight": "5px solid transparent",
                  "borderTop": 0,
                  "height": 0,
                  "width": 0,
                }
              }
            />
          </td>
          <td>
            my post
          </td>
        </tr>
      </table>
    </td>
  </tr>
  <tr>
    <td
      style={
        Object {
          "color": "#ff6600",
          "fontWeight": "bold",
          "padding": "5px",
        }
      }
    >
      <div
        style={
          Object {
            "float": "right",
          }
        }
      >
        <span>
          Previous 
        </span>
        |
        <span>
           Next
        </span>
      </div>
    </td>
  </tr>
  <tr>
    <td
      style={
        Object {
          "borderBottom": "4px solid #ff6600",
          "borderTop": "2px solid #ff6600",
        }
      }
    />
  </tr>
</table>
`;

exports[`app component render 1`] = `
<table>
  <tr>
    <td>
      <table>
        <tr
          bgcolor="#ff6600"
          style={
            Object {
              "color": "white",
              "fontWeight": "bold",
            }
          }
        >
          <td>
             Comments 
          </td>
          <td>
             Vote Count
          </td>
          <td>
             Upvote
          </td>
          <td>
             News Details
          </td>
        </tr>
      </table>
    </td>
  </tr>
  <tr>
    <td
      style={
        Object {
          "color": "#ff6600",
          "fontWeight": "bold",
          "padding": "5px",
        }
      }
    >
      <div
        style={
          Object {
            "float": "right",
          }
        }
      >
        <span>
          Previous 
        </span>
        |
        <span>
           Next
        </span>
      </div>
    </td>
  </tr>
  <tr>
    <td
      style={
        Object {
          "borderBottom": "4px solid #ff6600",
          "borderTop": "2px solid #ff6600",
        }
      }
    />
  </tr>
</table>
`;
